

syntax = "proto3";

package function_executor_service;

// The messages should not use any Python SDK objects. Only Function Executor implemented
// in Python is allowed to import Python SDK to run customer functions. This ensures that
// all the other components can be written in any language.

message SerializedObject {
    oneof data {
        // Set bytes_data if the object is serialized as bytes.
        bytes bytes = 1;
        // Set string_data if the object is serialized as string.
        string string = 2;
    }
    // The content type determines the serializer used to serialize the object.
    optional string content_type = 3;
}

// InitializeRequest contains information about the function
// that Function Executor is going to run the tasks for.
message InitializeRequest {
    optional string namespace = 1;
    optional string graph_name = 2;
    optional int32 graph_version = 3;
    optional string function_name = 5;
    optional SerializedObject graph = 7;
}

message InitializeResponse {
    optional bool success = 1;
}

message SetInvocationStateRequest {
    optional string key = 1;
    optional SerializedObject value = 2;
}

message SetInvocationStateResponse {}

message GetInvocationStateRequest {
    optional string key = 1;
}

message GetInvocationStateResponse {
    optional string key = 1;
    optional SerializedObject value = 2;
}

// InvocationStateRequest is sent by RPC Server to the client
// to perform actions on a task's graph invocation state.
message InvocationStateRequest {
    // The ID of the request sent by the client.
    // Must be unique per Function Executor.
    optional string request_id = 1;
    // The ID of the task initiated the request.
    optional string task_id = 2;
    oneof request {
        SetInvocationStateRequest set = 3;
        GetInvocationStateRequest get = 4;
    }
}

// InvocationStateResponse is sent by RPC client to the Server.
// A response contains the result of the action performed on the
// task's graph invocation state.
message InvocationStateResponse {
    // The id of the request this response is for.
    optional string request_id = 1;
    optional bool success = 2;
    oneof response {
        SetInvocationStateResponse set = 3;
        GetInvocationStateResponse get = 4;
    }
}

message FunctionOutput {
    repeated SerializedObject outputs = 1;
}

message RouterOutput {
    repeated string edges = 1;
}

message RunTaskRequest {
    optional string graph_invocation_id = 4;
    optional string task_id = 6;
    optional SerializedObject function_input = 9;
    optional SerializedObject function_init_value = 10;
}

message RunTaskResponse {
    optional string task_id = 1;
    optional FunctionOutput function_output = 2;
    optional RouterOutput router_output = 3;
    optional string stdout = 4;
    optional string stderr = 5;
    optional bool is_reducer = 6;
    optional bool success = 7;
}

service FunctionExecutor {
    // Initializes the Function Executor to run tasks
    // for a particular function. This method is called only
    // once per Function Executor as it can only run a single function.
    // It should be called before calling RunTask for the function.
    rpc initialize(InitializeRequest) returns (InitializeResponse);
    // Initializes a server that sends requests to the client to perform actions on
    // a task's graph invocation state. This method is called only once per Function Executor
    // It should be called before calling RunTask for the function.
    rpc initialize_invocation_state_server(stream InvocationStateResponse) returns (stream InvocationStateRequest);
    // Executes the task defined in the request.
    // Multiple tasks can be running in parallel.
    rpc run_task(RunTaskRequest) returns (RunTaskResponse);
}